kmk/keys.py

```diff
@@ -6,13 +11,20 @@
 from kmk.types import UnicodeModeKeyMeta
 from kmk.utils import Debug
 
+# Type aliases / forward declaration; can't use the proper types because of circular imports.
+Keyboard = object
+Key = object
```

Comment:
This sounds like the typing has caught a code smell then: those circular imports should be flattened down into a separate file imported in both places, perhaps.

Reply:
I don't see how. `KMKKeyboard` needs `Keys`, but `Keys` have `KMKKeyboard` as method argument.
Unless we make up dummy classes without implementation (which this more or less is), then that's an irreducible circular dependency.

Reply:
Thinking out loud: is this a case for what in TypeScript we'd call an "interface"? (maybe that's what you're getting at, actually).

```typescript
interface Thing {
  foo: string
  bar: number | undefined
  // I don't actually remember if this syntax is right right now
  buildOutput: () => string
}
```

And then anything that has those two properties and that method, with that signature, can be passed to something expecting a `Thing`

It's a layer of indirection in exchange for a layer of safety. In TypeScript, this stuff is all erased at compile time and has no runtime cost. Unsure what the Python equivalent would be, since ABCs have runtime componentry.

Reply:
That's a decent analogy, altough obviously pythons typehints only check for the name of the type, not for any implementation details (afaik).
There may be a tiny runtime impact: the type alias may be held in the modules dictionary; I'm not familiar enough with the CP parser and vm to give an educated answer.